hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java

```diff
@@ -3122,6 +3136,131 @@ public boolean setAcl(OzoneObj obj, List<OzoneAcl> acls) throws IOException {
+    long checkpointSnapshotIndex = omDBcheckpoint.getRatisSnapshotIndex();
+    if (checkpointSnapshotIndex <= lastAppliedIndex) {
+      LOG.error("Failed to install checkpoint from OM leader: {}. The last " +
+          "applied index: {} is greater than or equal to the checkpoint's " +
+          "snapshot index: {}", leaderId, lastAppliedIndex,
+          checkpointSnapshotIndex);
+      return null;
+    }
+
+    // Stop the ratis server so that no new transactions are applied. This
+    // can happen if a leader election happens while the state is being
+    // re-initialized.
+    omRatisServer.stop();
+
+    // Clear the OM Double Buffer so that if there are any pending
+    // transactions in the buffer, they are discarded.
+    omDoubleBuffer.stop();
+
+    // Take a backup of the current DB
+    File dbFile = metadataManager.getStore().getDbLocation();
+    String dbBackupFileName = OzoneConsts.OM_DB_BACKUP_PREFIX +
+        lastAppliedIndex + "_" + System.currentTimeMillis();
+    File dbBackupFile = new File(dbFile.getParentFile(), dbBackupFileName);
+
+    try {
+      Files.move(dbFile.toPath(), dbBackupFile.toPath());
+    } catch (IOException e) {
+      LOG.error("Failed to create a backup of the current DB. Aborting " +
+          "snapshot installation.", e);
+      return null;
+    }
+
+    // Move the downloaded DB checkpoint into the om metadata dir
+    Path checkpointPath = omDBcheckpoint.getCheckpointLocation();
+    try {
+      Files.move(checkpointPath, dbFile.toPath());
+    } catch (IOException e) {
+      LOG.error("Failed to move downloaded DB checkpoint {} to metadata " +
+          "directory {}",checkpointPath, dbFile.toPath(), e);
+      return null;
+    }
+
+    // Reload the OM DB store with the new checkpoint
+    try {
+      reloadOMState();
+    } catch (IOException e) {
+      LOG.error("Failed to reload OM state with new DB checkpoint.", e);
+      return null;
+    }
+
+    // TODO: We should only return the snpashotIndex to the leader.
+    //  Fixed after RATIS-586
+    TermIndex newTermIndex = TermIndex.newTermIndex(0,
+        checkpointSnapshotIndex);
+
+    return newTermIndex;
+  }
+
+  /**
+   * Re-instantiate MetadataManager with new DB checkpoint.
+   * All the classes which use/ store MetadataManager should also be updated
+   * with the new MetadataManager instance.
+   */
+  private void reloadOMState() throws IOException {
+
+    metadataManager = new OmMetadataManagerImpl(configuration);
+
+    metadataManager.start(configuration);
+
+    // Set metrics and start metrics back ground thread
+    metrics.setNumVolumes(metadataManager.countRowsInTable(metadataManager
+        .getVolumeTable()));
+    metrics.setNumBuckets(metadataManager.countRowsInTable(metadataManager
+        .getBucketTable()));
+
+    // Delete the omMetrics file if it exists
+    Files.deleteIfExists(getMetricsStorageFile().toPath());
+
+    // Re-initialize metadataManager dependent implementations
```

Comment:
Can some of this code be shared with startup initialization by moving to a common function?